src/librustc/traits/select.rs

```diff
@@ -1457,6 +1468,38 @@ impl<'cx, 'gcx, 'tcx> SelectionContext<'cx, 'gcx, 'tcx> {
         })
     }
 
+    fn assemble_generator_candidates(&mut self,
+                                   obligation: &TraitObligation<'tcx>,
+                                   candidates: &mut SelectionCandidateSet<'tcx>)
+                                   -> Result<(),SelectionError<'tcx>>
+    {
+        if self.tcx().lang_items.gen_trait() != Some(obligation.predicate.def_id()) {
+            return Ok(());
+        }
+
+        // ok to skip binder because the substs on generator types never
+        // touch bound regions, they just capture the in-scope
+        // type/region parameters
+        let self_ty = *obligation.self_ty().skip_binder();
+        let (closure_def_id, substs) = match self_ty.sty {
+            ty::TyGenerator(id, substs, _) => (id, substs),
+            ty::TyInfer(ty::TyVar(_)) => {
+                debug!("assemble_generator_candidates: ambiguous self-type");
+                candidates.ambiguous = true;
+                return Ok(());
+            }
+            _ => { return Ok(()); }
+        };
+
+        debug!("assemble_generator_candidates: self_ty={:?} obligation={:?}",
+               self_ty,
+               obligation);
+
+        candidates.vec.push(GeneratorCandidate(closure_def_id, substs));
```

Comment:
Why do we store the `closure_def_id` and `substs` in the candidate? We can *easily* recover them while confirming. I suppose the answer is "closures do that today" and I should ask @nikomatsakis why they do that.